runtime/src/bank.rs

```diff
@@ -2500,12 +2497,28 @@ impl Bank {
     }
 
     pub fn deposit(&self, pubkey: &Pubkey, lamports: u64) {
-        let mut account = self.get_account(pubkey).unwrap_or_default();
-        self.collected_rent.fetch_add(
-            self.rent_collector.update(pubkey, &mut account),
-            Ordering::Relaxed,
-        );
+        let account = self.get_account(pubkey);
+        let created = account.is_none();
+
+        let mut account = account.unwrap_or_default();
         account.lamports += lamports;
```

Comment:
@t-nelson Ugh, I've found some concerns while testing against local/testnet/mainnet-beta....

Specifically, the order of `deposit` and `rent` is significant in a corner case when the account is to be reclaimed, meaning there is too little lamports at the time of loading..

Also, there is some inconsistency for rent collection timing semantics with https://github.com/solana-labs/solana/issues/10426#issuecomment-640375379. Strictly abide by the rules, we would end up doing rent collection twice in deposit because we're both loading and storing at once. But I think the divergent behavior here is tolerable because `bank::deposit()` is fairly limited use.

I'm thinking for a good compromise... At bottom, I want to finish this without much complexities introduced and without any cluster consistency issue when updating.

Reply:
Ok, my thinking is making progress with a fresh brain after weekend.

Come to think of it, `update_rewards()` doesn't care about rents at all.

So, maybe `deposit()` should follow the suit as well? So, the general idea would be that **any kinds of protocol-level account manipulation is entirely exempt from rent manipulation.** That will simplify the reasoning a lot.

Even if we do that, the accounts will be rent-collected at least once an epoch by eager rent collection.

Reply:
I've checked the origin of rent in `deposit()`: https://github.com/solana-labs/solana/pull/6947/files#r347114185

As far as I understand, it seems that the reasoning is that all updates should be rent-collected. This is now realized by the eager rent collection now.

Also, I found another protocol-level account manipulation: collected rent distribution. That should be free of rent collection to avoid recursion..

So, I'm more inclined to the general consistent idea of no rent collection for protocol-level account manipulation.

Also, this would slightly alleviate rather hefty processing of `update_rewards` by avoiding rent collection there for the consistency of rent-collection coverage.